### ISSUE TITLE
Bumping MAX_BOTS value from 8 to 10, or anything greater than 8 causing immediate crash at startup

### ISSUE THREAD

**Issue by SirFroweey** [fgsfdsfgs/perfect_dark#226]
Bumping MAX_BOTS value from 8 to 10, or anything greater than 8 causing immediate crash at startup as shown below:

<img width="481" alt="image" src="https://github.com/fgsfdsfgs/perfect_dark/assets/1275689/db5f1349-20db-42e9-a2a2-c94b3bd4d911">

Interestingly the same code change (just changing MAX_BOTS to 10 from 8) works fine while on the perfect dark decompilation project repos master branch. 

I suspect this version of PD has some modifications to how bots are allocated in some sort of way?

Any pointer would be helpful as I am lost as to why its not working for this repository.

**Comment by fgsfdsfgs** [comment]
This is because mpconfigs and mpstrings are loaded from the ROM. They contain arrays whose sizes are `[MAX_BOTS]`. If you just increase the constant without changing those to `8`, the game will crash when trying to load them because in the ROM those arrays still have a length of 8.
Note that if you change the length to 8, the game will possibly read out of bounds when reading bot names and configs for the extra bots, so you might have to deal with that somehow.

**Comment by SirFroweey** [comment]
So how do I modify those from the ROM? Do I need to re-generate the ROM with those changes to mpconfgis/mstrings via the decompilation repo first? If so, what change needs to be done on this repository, would it be just the code (non-asset) changes? 

Also your response mentions:
> increase the constant without changing those to 8
and 
> Note that if you change the length to 8, the game will possibly read out of bounds when reading bot names and configs for the extra bots, so you might have to deal with that somehow.

Did you mean _from `8`_? Since `MAX_BOTS` is already `8` by default right

**Comment by SirFroweey** [comment]
@fgsfdsfgs 

So how do I modify those from the ROM? Do I need to re-generate the ROM with those changes to mpconfigs/mstrings via the decompilation repo first? If so, what change needs to be done on this repository, would it be just the code (non-asset) changes? 

Also your response mentions:
> increase the constant without changing those to 8

and

> Note that if you change the length to 8, the game will possibly read out of bounds when reading bot names and configs for the extra bots, so you might have to deal with that somehow.

Did you mean _from `8`_? Since `MAX_BOTS` is already `8` by default right

**Comment by fgsfdsfgs** [comment]
No, I meant change the `[MAX_BOTS]` in those two structs to literally `[8]`. Try doing that and see if it works fine.

**Comment by SirFroweey** [comment]
@fgsfdsfgs 

Git diff, to see if I changed them in the correct position:

<img width="394" alt="Screenshot 2023-11-05 131255" src="https://github.com/fgsfdsfgs/perfect_dark/assets/1275689/a2c77d3c-7487-4ab8-a833-a9b2b0e0ba6c">


After building/running I didn't crash which is great.

Forgive me, I'm a tad confused as to why changing this explicitly to `8` would "solve" the problem. Should it not be equal to the value I defined in `#define MAX_BOTS`, i.e.: `10`? And wouldn't `MAX_BOTS` default to `8` on the ROM anyways?

**Comment by fgsfdsfgs** [comment]
The ROM has `MAX_BOTS` set to `8`, so in there in `mpconfig` and `mpstrings` structures those arrays have size `8`. The port expects those structures to have the same size in both the ROM and the code, because it reads them from the ROM without modifying that.
A "proper" solution would be to separate these and convert from one to another when reading the ROM, but that requires a little bit of work to implement. This will likely be done when I get around to making the code 64-bit compatible, because in that case a lot of other structures will have different sizes too.

**Comment by SirFroweey** [comment]
@fgsfdsfgs 

Ah I think I understand now. Thank you so much for your patience. So basically, the change we just did is on the PORT to ensure it doesn't try to access data from the ROMs arrays that isn't there since the ROM has arrays for both structs as length of `8` where the change to `MAX_BOTS` I did on the port explicitly changes that to whatever value I assign, in this case `10`. So there's a data mismatch here and the change you recommended effectively "solves it".

Perhaps my understanding of the broader port project needs improvement. So when I do changes to the `src/` files this applies only to `pd.exe`? The ROM code remains unchanged right? I believed the entire game was repackaged as a windows PE and didn't know we still needed the ROM for data. How would I differentiate what pd.exe uses the ROM data for and what not?
 
Also could I change the ROMs code via the decompilation project and use it here to also "solve" the issue with the arrays not matching? Or is my head not in the right place?

**Comment by fgsfdsfgs** [comment]
>So when I do changes to the src/ files this applies only to pd.exe? The ROM code remains unchanged right?

Correct.

>I believed the entire game was repackaged as a windows PE and didn't know we still needed the ROM for data. How would I differentiate what pd.exe uses the ROM data for and what not?

For ease and legality of distribution and ease of modding, the port loads all assets from the ROM or external files. Assets being everything except the code, code being anything in `src/game`, `src/lib` and `src/textureconfig.c`
This means you can't make changes to text strings, animations, textures, setups, etc by editing their code or JSON files in the port. The JSON files are only used to generate some constants that are used in the code.

**Comment by SirFroweey** [comment]
@fgsfdsfgs 

Since I plan on changing strings and inserting new ones into the JSON files along with changing game logic (such as the case of adding more simulants) I should be working off of the original decompilation project and not this one? It seems this one is built around your modding feature?

**Comment by fgsfdsfgs** [comment]
It really depends on what you want to do. The only way to add more simulants in the port would be to modify the code, so you'd have to do that either way if you want that to work on the port.

**Comment by SirFroweey** [comment]
@fgsfdsfgs 

The thing is what I want to accomplish is insert new rows into the Simulant Selection UI part of the Combat Simulator. I've done this on the original decompilation project by inserting new values onto the `src/assets/ntsc-final/lang/mpmenu.json` file as shown below:

```
    {
        "id": "L_MPMENU_092_9",
        "en": "9:\n",
        "gb": "9:\n",
        "jp": "9:\n",
        "it": "9:\n",
        "fr": "9:\n",
        "de": "9:\n",
        "es": "9:\n"
    },
    {
        "id": "L_MPMENU_092_10",
        "en": "10:\n",
        "gb": "10:\n",
        "jp": "10:\n",
        "it": "10:\n",
        "fr": "10:\n",
        "de": "10:\n",
        "es": "10:\n"
    }
```

But from what you've told me that isn't possible here in its current state?

**Comment by fgsfdsfgs** [comment]
Since you're gonna be editing the code anyway, in the port you can do this with string literals without editing the JSON:
```
	{
		MENUITEMTYPE_SELECTABLE,
		8,
		MENUITEMFLAG_LITERAL_TEXT, // note this flag
		(uintptr_t)"9:",
		(uintptr_t)&mpMenuTextSimulantName,
		menuhandlerMpSimulantSlot,
	},
```

**Comment by SirFroweey** [comment]
@fgsfdsfgs 

Unfortunately that resulted in a crash as soon as I clicked Simulants on the Combat Simulator menu.

<img width="306" alt="Screenshot 2023-11-05 193816" src="https://github.com/fgsfdsfgs/perfect_dark/assets/1275689/4936bde9-eebd-41db-a30a-06fbac12dcb6">

As its a memory access exception I imagine its due to the ROM still believing there are `8` MAX_BOTS while the port has MAX_BOTS set as `10`.

**Comment by fgsfdsfgs** [comment]
Can you show me what exact changes you made?
